File: java/parser/src/main/java/gov/lbl/parser/parser/NameVisitor.java
package gov.lbl.parser.parser;

import java.util.List;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Name;
import gov.lbl.parser.domain.Name_part;
import static java.util.stream.Collectors.toList;
import java.util.ArrayList;

public class NameVisitor extends modelicaBaseVisitor<Name> {
    @Override
    public Name visitName(modelicaParser.NameContext ctx) {
        List<String> dots = ctx.SYMBOL_DOT()==null ? null : ctx.SYMBOL_DOT()
            .stream()
            .map(SYMBOL_DOT -> SYMBOL_DOT.getText())
            .collect(toList());
        List<String> identifiers = ctx.IDENT()
            .stream()
            .map(IDENT -> IDENT.getText())
            .collect(toList());

        List<Name_part> name_parts = new ArrayList<Name_part>();
        if (dots != null && identifiers != null) {
            if (identifiers.size() == dots.size()+1) {
                name_parts.add(new Name_part(false, identifiers.get(0)));
            }
            for (int i=1; i<identifiers.size(); i++) {
                name_parts.add(new Name_part(dots.get(i-1).equals("."), identifiers.get(i)));
            }
        }
        return new Name(name_parts);
    }
}

File: jsParser/parser/NameVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Name = require('../domain/Name')
const Name_part = require('../domain/Name_part')

class NameVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitName (ctx) {
    const dots = []
    const idents = []

    for (const dot of ctx.SYMBOL_DOT()) {
      dots.push(dot.getText())
    }

    for (const ident of ctx.IDENT()) {
      idents.push(ident.getText())
    }

    const name_parts = []
    if (idents.length == dots.length + 1) {
      name_parts.push(new Name_part.Name_part(false, idents[0]))
    } else if (idents.length == dots.length) {
      name_parts.push(new Name_part.Name_part(true, idents[0]))
    }

    for (let i = 1; i < idents.length; i++) {
      name_parts.push(new Name_part.Name_part(dots[i - 1] == '.', idents[i]))
    }

    return new Name.Name(name_parts)
  }
};

NameVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitName = this.visitName
exports.NameVisitor = NameVisitor
